### PR TITLE
Feat/java wrapper

### PR DESCRIPTION
**Pull request by ArlonTM**
- Implemented routine to get the token from hosted web session
- Performed mild code cleanup

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -139,4 +139,11 @@ gradle-app.setting
...
-# End of https://www.gitignore.io/api/macos,gradle,intellij
\ No newline at end of file
+# End of https://www.gitignore.io/api/macos,gradle,intellij
+/bin/
+
+# Avoid sending Eclipse-generated files
+.classpath
+.project
+.settings/
+EntryPoint.java
```
Comment (JNdhlovu) [inline .gitignore:149]: Whats the purpose of this guy do we have a commited file we are now removing? or something you are using locally?

File: build.gradle [hunk 2]
```diff
@@ -86,13 +85,13 @@ uploadArchives {
 
                 developers {
                     developer {
-                        id = 'Ridhwana'
-                        name = 'Ridhwana Khan'
-                        email = 'ridhwana@smileidentity.com'
+                        id = 'Arlon'
```
Comment (JNdhlovu) [inline build.gradle:88]: Let's put support details here, including support@smileidentity.com this is not critical to the deploy but is metadata that people can pick

File: src/main/java/smile/identity/core/Signature.java [hunk 3]
```diff
@@ -72,11 +140,12 @@ public Boolean confirm_sec_key(String timestamp, String sec_key) throws Exceptio
         } catch (Exception e) {
             throw e;
         }
-
     }
 
     private static PublicKey loadPublicKey(String apiKey) throws GeneralSecurityException, IOException {
         apiKey = apiKey.replace("-----BEGIN PUBLIC KEY-----", "").replace("-----END PUBLIC KEY-----", "").replace("\n", "").replace("\r", "").trim();
+        apiKey = apiKey.replaceAll(" ", "");
+        System.out.println("API_KEY: " + apiKey);
```
Comment (JNdhlovu) [inline src/main/java/smile/identity/core/Signature.java:148]: Please remove logs

File: src/test/java/smile/identity/core/SignatureTest.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.Date;
+
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.junit.Before;
+import org.junit.Test;
+
+public class SignatureTest {
+	
+    private Integer PARTNER_ID = 212;
```
Comment (JNdhlovu) [inline src/test/java/smile/identity/core/SignatureTest.java:16]: This is partner sensitive partner info here and should not be commited please remove from git history as this is a public repo

**Review by JNdhlovu — CHANGES_REQUESTED** [review]
Please see comments

File: src/test/java/smile/identity/core/SignatureTest.java (hunk 4 above)
Comment (ArlonTM) [reply src/test/java/smile/identity/core/SignatureTest.java:16]: Valid observation! 👍🏽

File: src/main/java/smile/identity/core/Signature.java (hunk 3 above)
Comment (ArlonTM) [reply src/main/java/smile/identity/core/Signature.java:148]: 👍🏽

File: .gitignore (hunk 1 above)
Comment (ArlonTM) [reply .gitignore:149]: Either environment-specific files or testing classes, used locally.

File: README.md [hunk 5]
```diff
@@ -316,6 +319,27 @@ Please note that if you do not need to pass through Options if you will not be u
 String response = connection.get_job_status(partnerParameters.get(), null);
 ```
 
+##### get_hosted_web_session method
```
Comment (beaglebets) [inline README.md:322]: As this is the first implementation of this feature I think we should take a moment to think about the naming . Is this what we want to call it? @camerondgray your thoughts?

File: src/main/java/smile/identity/core/IDApi.java [hunk 6]
```diff
@@ -27,22 +29,17 @@ public IDApi(String partner_id, String api_key, Integer sid_server) throws Excep
...
-            if (sid_server.equals("0")) {
-                url = "https://3eydmgh10d.execute-api.us-west-2.amazonaws.com/test";
-            } else if (sid_server.equals("1")) {
-                url = "https://la7am6gdm8.execute-api.us-west-2.amazonaws.com/prod";
-            } else {
-                url = sid_server;
-            }
-        } catch (Exception e) {
-            throw e;
+    	this.partner_id = partner_id;
```
Comment (beaglebets) [inline src/main/java/smile/identity/core/IDApi.java:33]: looks like something is not indented correctly here.

File: src/main/java/smile/identity/core/WebApi.java [hunk 7]
```diff
@@ -79,99 +77,111 @@ public String submit_job(String partner_params, String images_params, String id_
...
+     * @param timestamp the timestamp to generate the token from
+     * @param user_id
+     * @param job_id
+     * @param job_type
+     * @param product_type
+     * @param signature a previously generated signature
+     * @return TRUE or FALSE
+     */
+    @SuppressWarnings("unchecked")
+	public String get_hosted_web_session(Long timestamp, String user_id, String job_id, int job_type, String product_type, String signature) throws Exception {
```
Comment (beaglebets) [inline src/main/java/smile/identity/core/WebApi.java:158]: It doesn't make sense to require the developer calling this function to pass in a timestamp and signature. This should automatically generate the timestamp and signature and add it to the request, like we do for the submit_job method.

**Review by beaglebets — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (camerondgray) [reply README.md:322]: Good call to stop and be thoughtful about this. My instinct would be to call it `get_web_token` and update the rest of the comments to be consistent around that. I am definitely open to other suggestions though.

Comment (ArlonTM) [reply README.md:322]: `get_web_token` makes more sense.

File: src/main/java/smile/identity/core/WebApi.java (hunk 7 above)
Comment (ArlonTM) [reply src/main/java/smile/identity/core/WebApi.java:158]: Totally right!

**Review by camerondgray — APPROVED** [review]

File: README.md [hunk 8]
```diff
@@ -318,6 +319,25 @@ Please note that if you do not need to pass through Options if you will not be u
 String response = connection.get_job_status(partnerParameters.get(), null);
 ```
 
+##### get_web_token method
+
+This function provides a wrapper around a hosted web session; it returns a stringified JSONObject containing an active web token
```
Comment (beaglebets) [inline README.md:324]:
```suggestion
This function provides the token to be used for authentication when using our hosted web session; it returns a stringified JSONObject containing an active web token.
```

File: src/test/java/smile/identity/core/WebApiTest.java [hunk 9]
```diff
@@ -3,12 +3,53 @@
...
+
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
 
 public class WebApiTest {
-    @Test public void testSomeLibraryMethod() {
-        // WebApi classUnderTest = new WebApi();
-        // assertTrue("someLibraryMethod should return 'true'", classUnderTest.someLibraryMethod());
+	
+	private static final int PORT = 32100; //Random port number, arbitrarily chosen
```
Comment (beaglebets) [inline src/test/java/smile/identity/core/WebApiTest.java:21]: The indentation in this file is inconsistent starting on this line and through the rest of the file. I recommend cleaning this up.

**Review by beaglebets — COMMENTED** [review]
I left a couple of comments that are not substantive, but stylistic. I will not hold up approval for them to be adressed.

File: src/test/java/smile/identity/core/WebApiTest.java (hunk 9 above)
Comment (ArlonTM) [reply src/test/java/smile/identity/core/WebApiTest.java:21]: Quite weird... 🤔

**Review by beaglebets — APPROVED** [review]

**Pull request merged**